compiler/rustc_mir/src/transform/coverage/spans.rs

```diff
@@ -246,8 +246,8 @@ impl<'a, 'tcx> CoverageSpans<'a, 'tcx> {
     ) -> Vec<CoverageSpan> {
         let mut coverage_spans = CoverageSpans {
             mir_body,
-            fn_sig_span,
-            body_span,
+            fn_sig_span: fn_sig_span.with_ctxt(SyntaxContext::root()),
```

Comment:
It doesn't look like the `SyntaxContext` is adjusted in any other places (e.g. in `bcb_to_initial_coverage_spans`). Is there something special about async desugarings that makes this necessary? What happens if the body of a function is generated from a macro (and therefore has a non-root `SyntaxContext`)?

Reply:
See `function_source_span()`, which does adjust the `SyntaxtContext` for all other spans.

All of the function body spans are extrapolated to their `rustc_span::source_map::original_sp`. 

They are assumed to then be in the context of the body. 

The bug, fixed by this PR, was that I assumed the body would always be in the root. Now all contexts are normalized to `root`. 

This allows me to use `Span::to()`, which, otherwise, doesn't always return the expected result, if the contexts are different (even though they are in the same source).

The end result (in tests and in many practical use cases) provide the evidence that this approach works, except for the bug fixed by this PR, which is now obvious.

Reply:
I'm taking a closer look at this though, and trying an alternative that may be relevant to the concern you raised.

Reply:
I just uploaded a modified solution (see the new commit). This still forces all spans (fn_sig, body, and the MIR sources) to the same ctxt, so `Span::to()` should still work, but it does not force everything to the root ctxt anymore. Instead, the `body_span`'s ctxt is now the common ctxt.

In the previous commit, I normalized the `body_span`'s ctxt to the root, but that would have changed the behavior or `original_sp`, when using the `body_span` as the enclosing span. That was the wrong change, and I'm glad to have had the opportunity to review this.